src/libs/elektra/key.c

```diff
@@ -362,32 +362,54 @@ int f (const Key * source)
  * @code
 int g (const Key * source, KeySet * ks)
 {
-	Key * dup = keyDup (source);
+	Key * dup = keyDup (source, 0);
 	// work with duplicate
 	ksAppendKey (ks, dup);
 	// ksDel(ks) will also free the duplicate
 	// source remains unchanged.
 }
  * @endcode
  *
- * Duplication of keys should be preferred to keyNew(),
- * because data like owner can be filled with a copy
- * of the key instead of asking the environment.
- * It can also be optimized in the checks, because the keyname
- * is known to be valid.
+ * Depending on what @p flags are given, only parts of the key will be copied.
+ * The supported #elektraKeyFlags values are:
+ *   - #KEY_NAME   - copies the name
+ *   - #KEY_VALUE  - copies the value
+ *   - #KEY_META   - copies all meta keys
+ *   - #KEY_BINARY - creates a binary key, even if @p source is not binary
+ *
+ * If #KEY_NAME is set, the returned Key will have the same name as @p source.
+ * Otherwise, the name will be `"/"`.
+ *
+ * If #KEY_VALUE is set, the returned Key will have the same value as @p source.*
+ * If #KEY_BINARY is set as well, the returned Key will be a binary key independent of the type of @p source.
+ *
+ * Without #KEY_VALUE, the returned Key will have an empty value. If #KEY_BINARY is set,
+ * the empty value is `NULL`, otherwise it is an emtpy string `""`.
+ *
+ * If #KEY_META is set, meta keys will be copied from @p source to the returned key.
+ * This function creates a deep copy of meta keys, i.e. all meta keys on @p source will be `keyDup`ed before
+ * adding them to the returned Key.
+ * If you do not need this, you should prefer:
+ * @code
+Key * dup = keyDup (source, 0);
+keyCopyAllMeta (dup, source);
+ * @endcode
+ *
+ * Calling `keyDup (source, 0)` is equivalent to `keyDup (source, KEY_NAME | KEY_VALUE)`.
```

Comment:
I find it very odd that a have to tell a function called `keyDup` (or even weirder `keyCopy`) what **not** to duplicate/copy. It seems much more intuitive to specify what should happen and not what shouldn't happen.

There is another reason for this API decision: Actually allowing `keyDup (source, 0)` wouldn't make sense. It would just be the same as `keyNew ("/", KEY_END)`.

Copying name and value also seemed like the best default, since you can use `keyDup (source, ~0)` to copy everything (even if we add new flags). In most situations, you will either want to copy just the name (needed once `keyNew` changes), everything. That would be `keyDup (source, KEY_NAME)` and `keyDup (source, ~0)` respectively. That leaves the somewhat rarer case of copying just name and value and `keyDup (source, 0)` makes that easy as well.

I think both making `keyDup (source, 0)` an error (i.e. returns `NULL`) and equivalent to `keyNew ("/", KEY_END)` is just a bit to error prone. Especially since other "flags" APIs in Elektra (e.g. `keyLookupByName`) also choose reasonable defaults if `0` is used. (of course over there some flags are positive and some negative, which makes the default a bit more explicit)

Reply:
> I find it very odd that a have to tell a function called keyDup (or even weirder keyCopy) what not to duplicate/copy. It seems much more intuitive to specify what should happen and not what shouldn't happen.

I could get used to either. It mostly depends on how the function&parameter is called if it is intuitive or not. The most important difference is what `keyDup(, 0)` should do, as this is the shortest to write and to get this consistent with anything else you can do.

I see three consistent ways:

1. make 0 copy everything and substract with flags
2. make 0 copy nothing and add with flags (KEY_ALL could be used to copy everything)
3. make 0 copy what we found reasonable, and have flags which add and substract, and the flag name makes it clear what happens.

Copying everything by default is reasonable for me, I would expect this from a duplication.

Reply:
> Copying everything by default is reasonable for me, I would expect this from a duplication.

Theoretically yes, but copying everything creates a deep copy of metadata and that could be quite expensive. I forgot to add that this was also a reason why `KEY_NAME | KEY_VALUE` is the default.

I would also rule out Option 3, because we would need new flags for that. It would also be pretty weird if e.g `keyDup (k, KEY_META)` copies everything, because `KEY_NAME` and `KEY_VALUE` have to be disabled. The only way I see this working is, if only `KEY_NAME` is active by default. It's still a bit weird for `keyCopy`, but `keyDup` without the name rarely makes sense.

That basically leaves Option 2: 0 is "copy nothing". It's also weird because it turns `keyDup (k, 0)` into `keyNew ("/", KEY_END)` and `keyCopy (dest, source, 0)` into `keyClear (dest)`. But I that's more of an odd side effect than a real downside.

Reply:
> Theoretically yes, but copying everything creates a deep copy of metadata and that could be quite expensive.

I do not understand why you want it to be a deep copy.

Reply:
> Why? It looks like a normal filter to me.

To me `keyCopy (dest, source, KEY_NAME)` reads very much as: "Please copy the name from source to dest".
It just seems more natural to me to tell a copy function what it should copy. For `keyDup` it is a bit better with the filter, but they should have the same behaviour.

But a better reason may be forward compatibility. If we add some new data `foo` to the key and introduce a new flag `KEY_FOO`. Existing `keyCopy` calls will automatically copy `foo` as well, because the `KEY_FOO` filter wasn't applied (because it didn't exist).

> I do not understand why you want it to be a deep copy.

See [above](https://github.com/ElektraInitiative/libelektra/pull/3606#discussion_r547872400), why a deep copy of metadata may be needed.

Reply:
I agree, that users might not expect the same semantics of the flags in keyDup vs. keyCopy and this is actually quite problematic.

Probably, we should not provide both dup+copy. Actually only copy is needed in the core API, dup can quite easily implemented by anyone as needed (a keyNew+keyCopy is not so much different to what our keyDup provides anyway).

The question now is if we can get rid of all occurrences in keyDup in the core.

Reply:
The problem is that `keyCopy` is more powerful (can overwrite existing keys), while `keyDup` is much more useful. For example, you can do things like `ksAppendKey (ks, keyDup (k, flags))`.

A simple fix would be to change the signature of `keyCopy` to

```c
Key * keyCopy (Key * dest, Key * source, elektraKeyFlags flags);
```

and return `dest`. For the error case, returning `NULL` could be a solution. The only two possible error cases are `dest == NULL` (would return `NULL` anyway) and `dest` is read-only (can be avoided by the programmer).

Then you could replace `keyDup (k, flags)` with `keyCopy (keyNew ("/", KEY_END), k, flags)`. 

> The question now is if we can get rid of all occurrences in keyDup in the core.

I have to look through all uses of `keyDup` anyway to set the correct flags, so replacing it shouldn't be that big of a problem.

Reply:
I like this idea very much!